ir/function.h

```diff
@@ -212,4 +213,43 @@ class DomTree final {
     void printDot(std::ostream &os) const;
 };
 
+class DepthFirstSpanningTree final {
+  Function &f;
+  CFG &cfg;
+  unsigned current = 0;
+  std::map<const BasicBlock *, unsigned> number;
+  std::vector<const BasicBlock *> node;
+  std::map<unsigned, unsigned> last;
+  std::map<const BasicBlock *, std::vector<const BasicBlock *>> edges;
+  void dfs (const BasicBlock *bb);
+  void getTree();
+public:
+  DepthFirstSpanningTree(Function &f, CFG &cfg) : f(f), cfg(cfg) { getTree(); }
+  const std::map<unsigned, unsigned>& get_last() const { return last; };
+  const std::map<const BasicBlock *, unsigned>& get_number() const { return number; };
+  const std::vector<const BasicBlock *>& get_node() const { return node; };
+};
+
+class LoopAnalysis final {
+  Function &f;
+  CFG cfg;
+  DepthFirstSpanningTree dfst;
+  std::vector<unsigned> header;
+  enum NodeType { nonheader, self, reducible };
+  std::vector<NodeType> type;
+  void printNodeType(std::ostream &os, const NodeType &t) const;
+  void analysis();
+public:
+  LoopAnalysis(Function &f) : f(f), cfg(f), dfst(f, cfg) { analysis(); }
+  void printDot(std::ostream &os) const;
+};
+
+class DisjointSet final {
```

Comment:
this is a generic datastructure. should be moved to utils.
Also, is this a union-find? if so, please change the name to make that explicit.